src/plugins/mer/merqmakebuildconfigurationfactory.cpp

```diff
@@ -0,0 +1,184 @@
+        m_MerSsh.close();
+    });
+
+    connect(project, &QmakeProject::proFileUpdated,
+            [this, project](QmakeProjectManager::QmakeProFile *pro, bool validParse, bool parseInProgress){
+        if (project == pro->project() && validParse == true && parseInProgress == true)
+            queryBuildEngineVariables();
+    });
+}
+
+
+
+QStringList MerQmakeBuildConfiguration::merSshEnvironment() const
+{
+    if (!target() || !target()->project() || !target()->kit())
+        return QStringList();
+
+    const auto kit = target()->kit();
+    const auto sdk = MerSdkKitInformation::sdk(kit);
+    const auto qtVersion = QtSupport::QtKitInformation::qtVersion(kit);
+    if (!sdk || !qtVersion)
+        return QStringList();
+
+    const QString sshPort = QString::number(sdk->sshPort());
+    const QString sharedHome = QDir::fromNativeSeparators(sdk->sharedHomePath());
+    const QString sharedTarget = QDir::fromNativeSeparators(sdk->sharedTargetsPath());
+    const QString sharedSrc = QDir::fromNativeSeparators(sdk->sharedSrcPath());
+    const auto item = [](const QString &name, const QString &value)->QString {
+        Q_ASSERT(!name.contains(QChar::Space));
+        return QString("%1=%2").
+                arg(name).
+                arg(value.contains(QChar::Space) ?
+                        "\"" + value + "\"" :
+                        value);
+    };
+
+    auto sysenv = QProcess::systemEnvironment();
+    sysenv << item(Constants::MER_SSH_USERNAME, sdk->userName()) <<
+              item(Constants::MER_SSH_PORT, sshPort) <<
+              item(Constants::MER_SSH_PRIVATE_KEY, sdk->privateKeyFile()) <<
+              item(Constants::MER_SSH_SHARED_HOME, sharedHome) <<
+              item(Constants::MER_SSH_SHARED_TARGET, sharedTarget) <<
+              item(Constants::MER_SSH_TARGET_NAME, MerSdkManager::targetNameForKit(kit)) <<
+              item(Constants::MER_SSH_PROJECT_PATH, target()->project()->projectDirectory().toString()) <<
+              item(Constants::MER_SSH_SDK_TOOLS, qtVersion->qmakeCommand().parentDir().toString());
+    if (!sharedSrc.isEmpty())
+        sysenv << item(Constants::MER_SSH_SHARED_SRC, sharedSrc);
+
+    return sysenv;
+}
+
+
+
+void MerQmakeBuildConfiguration::onMerSshFinished()
+{
+    if (isFinished(m_MerSsh)) {
+        const auto buildEngineVariables = readFrom(m_MerSsh);
+        if (buildEngineVariables != m_BuildEngineVariables) {
+            m_BuildEngineVariables = buildEngineVariables;
+            updateCacheAndEmitEnvironmentChanged();
+        }
+    }
+    m_MerSsh.close();
+}
+
+
+
+void MerQmakeBuildConfiguration::queryBuildEngineVariables()
+{
+    if (m_MerSsh.state() == QProcess::NotRunning) {
+        m_MerSsh.setEnvironment(merSshEnvironment());
+        m_MerSsh.start();
+    }
+}
+
+
+
+void MerQmakeBuildConfiguration::addToEnvironment(Utils::Environment &env) const
+{    
+    foreach (const QString &var, m_BuildEngineVariables) {
```

Comment:
Copying unfiltered Qmake variables from build engine to the process environment may have undesired effects.